manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,1033 @@
+		select {
+		case <-ctx.Done():
+			log.G(ctx).Debug("context done, canceling the event stream")
+			cancel()
+		}
+	}()
+
+	// now restore the local state
+	log.G(ctx).Info("restoring local network allocator state")
+	if err := a.network.Restore(networks, services, tasks, nodes); err != nil {
+		log.G(ctx).WithError(err).Error("error restoring local network allocator state")
+		// if restore fails, we cannot proceed with the allocator. we need to
+		// stop it here, so that we don't perform work and get into a bad
+		// state.
+		return err
+	}
+
+	// allocate all of the store in its current state. we do this becauses at
+	// this level, we have no idea which objects are allocated or not; only the
+	// network package and below knows. before we can start allocating new
+	// stuff off the event stream, however, we need to allocate any old stuff
+	// that wasn't allocated before we started. so, we add every object to the
+	// pending map. those that are already allocated will simply return
+	// ErrAlreadyAllocated when we try to allocate them, which we can ignore,
+	// and those that aren't will be. after this first pass, we'll be in a
+	// consistent state ready to start processing new allocations.
+	for _, network := range networks {
+		a.pendingNetworks[network.ID] = struct{}{}
+	}
+	for _, node := range nodes {
+		a.pendingNodes[node.ID] = struct{}{}
+	}
+	for _, task := range tasks {
+		a.pendingTasks[task.ID] = struct{}{}
+	}
+
+	log.G(ctx).Info("processing all outstanding allocations in the store")
+	a.processPendingAllocations(ctx)
+	log.G(ctx).Info("finished processing all pending allocations")
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	log.G(ctx).Debug("starting event watch loop")
+	// this goroutine handles incoming store events. all we need from the
+	// events is the ID of what has been updated. by the time we service the
+	// allocation, the object may have changed, so we don't save any other
+	// information. we'll get to it later.
+	go func() {
+		// shadow context. trivial optimization to avoid calling this in every
+		// single event handle body
+		ctx := log.WithField(ctx, "method", "(*NewAllocator).handleEvent")
+		defer wg.Done()
+		// TODO(dperny): docker/swarmkit#2619: there is an outstanding issue
+		// with a deadlock that can result from closing the store at the same
+		// time the callback watch is canceled. This can cause the tests to
+		// deadlock. It may also be a concern during node demotion.
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case event := <-watch:
+				a.handleEvent(ctx, event)
+			}
+		}
+	}()
+
+	log.G(ctx).Debug("starting batch processing loop")
+	// allocations every 500 milliseconds
+	batchTimer := time.NewTimer(maxBatchInterval)
+	go func() {
+		defer wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-batchTimer.C:
+				a.pendingMu.Lock()
+				total := len(a.pendingNetworks) + len(a.pendingNodes) + len(a.pendingTasks)
+				a.pendingMu.Unlock()
```

Comment:
so, basically, as a future improvement, we will want to make `processPendingAllocations` more intelligent. for example, right now, if allocation fails because resources are exhausted, we will start spinning and trying to allocate. ideally we would in that case want to check if something has been deallocated before launching into another attempt to allocate. (this was actually the case in the old allocator, which I chose to ignore in the interest of time and simplicity for this PR). the check for this would happen here in this function, and NOT in processPendingAllocations, which should have the sole responsibility of doing the actual work, not deciding if it should do the work.

does that make sense?

Reply:
If we want to check whether something has been deallocated, we might have to lock here too indeed ; however I'm not sure it's worth locking twice for now, we don't know what we'll actually do in the future and when. Also that check in `processPendingAllocations` can't hurt I think, and it would get rid of this lock for now. All in all I think the PR should focus on making it sensible as is, not leave a placeholder for later - but I'll let you judge.